README.md

```diff
@@ -1,21 +1,77 @@
 # golf-course
 A list of common Solidity optimization tips and myths.
 
-## Tips
+# Tips
 
-### Right Shift Instead of Dividing By 2
+- - - -
+### 1.  When dividing by two, use `>> 1` instead of `/ 2` ###
+
+```solidity
+/// 🤦 Unoptimized (gas: 1401)
+uint256 two = 4 / 2;
+
+/// 🚀 Optimized (gas: 1372)
+uint256 two = 4 >> 1;
+```
 
 The `SHR` opcode is 3 gas cheaper than `DIV` and also bypasses Solidity's division by 0 prevention overhead.
+  - [Gas Usage]()
+  - [Full Example](https://github.com/Rari-Capital/golf-course/blob/fc1882bacfec50787d9e9435d59fed4a9091fb21/src/GolfCourse.t.sol#L15)
+
+
+- - - -
+
+### 2.  Use `>=` and `<=` instead of `>` and `<` ###
+```solidity
+/// 🤦 Unoptimized (gas: 1401)
+bool positive = 1 > 0;
+
+/// 🚀 Optimized (gas: 1337)
+bool positive = 1 >= 1;
+```
+
+`if (x >= y) {}` is algebraically equivalent to !(x < y) which gets compiled to `LT .. JUMPI` which is more efficient than `x > y` which contains an extra `ISZERO`
+
+- [Gas Usage]()
+- [Full Example](https://github.com/Rari-Capital/golf-course/blob/fc1882bacfec50787d9e9435d59fed4a9091fb21/src/GolfCourse.t.sol#L23)
+
+- - - -
+
+### 3. Using `!=` is usually cheaper than `>` or `<` ###
+```solidity
+/// 🤦 Unoptimized (gas: 1423)
+notThirtyTwo = 31 < 32;
+
+/// 🚀 Optimized (gas: 1426)
+notThirtyTwo = 31 != 32;
+```
+Similar to `>=` and `<=` - the `!=` gets compiled without an extra ISZERO in many cases (One notable exception being `x != 0` gets compiled the same as `x > 0`)
 
 - [Gas Usage]()
-- [Full Example]()
+- [Full Example](https://github.com/Rari-Capital/golf-course/blob/fc1882bacfec50787d9e9435d59fed4a9091fb21/src/GolfCourse.t.sol#L31)
+
 
+
+- - - -
+
+### 4. Use `<address>.code.length` instead of assembly `extcodesize` to avoid variable overhead ###
 ```solidity
-// Unoptimized:
-uint256 two = 4 / 2;
+/// 🤦 Unoptimized (gas: 1535)
```

Comment:
In spite of the increased variable overhead, the gas is still cheaper using the assembly.  This kinda goes against our idea of putting in controls to ensure that every "optimization" is less gas than the unoptimized version.  Wdyt?  Perhaps we could have separate section for optimizing for variable overhead?

Reply:
very possible its not a saving technique on its on, the savings do come in tho if u're accessing codesize via a library like Address.sol which has function call overhead associated. but yeah maybe not good for this guide then

Reply:
removing it from tips, we'll put this on the back burner